src/plugins/dashboard/public/application/hooks/use_dashboard_app_state.ts

```diff
@@ -262,9 +262,13 @@ export const useDashboardAppState = ({
         dashboardBuildContext.$checkForUnsavedChanges,
       ])
         .pipe(debounceTime(DashboardConstants.CHANGE_CHECK_DEBOUNCE))
-        .subscribe((states) => {
+        .subscribe(async (states) => {
```

Comment:
Good call @Dosant ! 
I didn't think of it, because the function in its current form is almost synchronous due to the only async piece being the `untilEmbeddableLoaded` method which is synchronous unless the dashboard is loading.

That said, this could have easily caused race conditions if a downstream implementation used a long-running task in their diffing method. I added a switchmap operator to the pipe and return when the inner observable is closed. I tested by adding a one second delay to the `diff_dashboard_state` function, firing off many changes in quick succession, and ensuring that only the latest changes are added to the session storage.

Reply:
nit: maybe it can be simpler with wrapping promise into an observable: 

```
stream$.pipe(switchMap((v) => from(op(v)))).subscribe((v) => {
 // logic
});
```